warehouse/query-core/src/main/java/datawave/query/function/LimitFields.java

```diff
@@ -87,89 +88,127 @@ public Multimap<String,Attribute<? extends Comparable<?>>> load(String key) {
             if (this.limitFieldsMap.containsKey(keyNoGrouping)) { // look for the key without the grouping context
                 if (log.isTraceEnabled())
                     log.trace("limitFieldsMap contains " + keyNoGrouping);
+                
                 Attribute<?> attr = de.getValue();
-                int limit = this.limitFieldsMap.get(keyNoGrouping); // used below if you un-comment to get all hits
                 if (attr instanceof Attributes) {
                     Attributes attrs = (Attributes) attr;
                     Set<Attribute<? extends Comparable<?>>> attrSet = attrs.getAttributes();
                     for (Attribute<? extends Comparable<?>> value : attrSet) {
-                        manageHitsAndMisses(keyWithGrouping, keyNoGrouping, value, hitTermMap, hits, misses, countForFieldMap);
+                        if (isHit(keyWithGrouping, value, hitTermMap)) {
+                            increment(keyNoGrouping, countKeepersForFieldMap);
+                        } else {
+                            value.setToKeep(false);
+                            increment(keyNoGrouping, countMissesRemainingForFieldMap);
+                        }
+                        increment(keyNoGrouping, countForFieldMap);
                     }
-                    
                 } else {
-                    manageHitsAndMisses(keyWithGrouping, keyNoGrouping, attr, hitTermMap, hits, misses, countForFieldMap);
+                    if (isHit(keyWithGrouping, attr, hitTermMap)) {
+                        increment(keyNoGrouping, countKeepersForFieldMap);
+                    } else {
+                        attr.setToKeep(false);
+                        increment(keyNoGrouping, countMissesRemainingForFieldMap);
+                    }
+                    increment(keyNoGrouping, countForFieldMap);
                 }
             }
         }
-        for (String keyNoGrouping : countForFieldMap.keySet()) {
-            int limit = this.limitFieldsMap.get(keyNoGrouping);
-            Multimap<String,Attribute<? extends Comparable<?>>> hitMap = hits.getUnchecked(keyNoGrouping);
-            for (String keyWithGrouping : hitMap.keySet()) {
-                for (Attribute<? extends Comparable<?>> value : hitMap.get(keyWithGrouping)) {
-                    // if(limit <= 0) break; // comment this line if you want to get ALL hits even if the limit is exceeded
-                    reducedMap.put(keyWithGrouping, value);
-                    limit--;
-                }
+        
+        // second pass is to set any misses back to be kept if the limit allows
+        for (Map.Entry<String,Attribute<? extends Comparable<?>>> de : document.entrySet()) {
+            String keyWithGrouping = de.getKey();
+            String keyNoGrouping = keyWithGrouping;
+            // if we have grouping context on, remove the grouping context
+            if (keyNoGrouping.indexOf('.') != -1) {
```

Comment:
The computing of the key with no grouping also appears above. We could make a function and reuse:

```
String keyNoGrouping = removeGrouping(keyWithGrouping);
```

Reply:
good one.  will do